core/shared/src/main/scala/zio/RefM.scala

```diff
@@ -1,234 +1,22 @@
+   * @see [[zio.ZRefM.makeManaged]]
    */
-  def update[R, E](f: A => ZIO[R, E, A]): ZIO[R, E, Unit] =
-    modify(a => f(a).map(a => ((), a)))
+  def makeManaged[A](a: A): UManaged[RefM[A]] =
+    ZRefM.makeManaged(a)
 
   /**
-   * Atomically modifies the `RefM` with the specified function, returning the
-   * value immediately after modification.
-   *
-   * @param f function to atomically modify the `RefM`
-   * @tparam R environment of the effect
-   * @tparam E error type
-   * @return `ZIO[R, E, A]` modified value of the `RefM`
+   * @see [[zio.ZRefM.signalRef]]
    */
-  def updateAndGet[R, E](f: A => ZIO[R, E, A]): ZIO[R, E, A] =
-    modify(a => f(a).map(a => (a, a)))
-
-  /**
-   * Atomically modifies the `RefM` with the specified partial function.
-   * If the function is undefined on the current value it doesn't change it.
-   *
-   * @param pf partial function to atomically modify the `RefM`
-   * @tparam R environment of the effect
-   * @tparam E error type
-   * @return `ZIO[R, E, Unit]`
-   */
-  def updateSome[R, E](pf: PartialFunction[A, ZIO[R, E, A]]): ZIO[R, E, Unit] =
-    modify(a => pf.applyOrElse(a, (_: A) => IO.succeedNow(a)).map(a => ((), a)))
-
-  /**
-   * Atomically modifies the `RefM` with the specified partial function.
-   * If the function is undefined on the current value it returns the old value
-   * without changing it.
-   *
-   * @param pf partial function to atomically modify the `RefM`
-   * @tparam R environment of the effect
-   * @tparam E error type
-   * @return `ZIO[R, E, A]` modified value of the `RefM`
-   */
-  def updateSomeAndGet[R, E](pf: PartialFunction[A, ZIO[R, E, A]]): ZIO[R, E, A] =
-    modify(a => pf.applyOrElse(a, (_: A) => IO.succeedNow(a)).map(a => (a, a)))
-}
-
-object RefM extends Serializable {
-  private[RefM] final case class Bundle[E, A, B](
-    interrupted: Ref[Option[Cause[Nothing]]],
-    update: A => IO[E, (B, A)],
-    promise: Promise[E, B]
-  ) {
-    def run(a: A, ref: Ref[A], onDefect: Cause[E] => UIO[Unit]): UIO[Unit] =
-      interrupted.get.flatMap {
-        case Some(cause) => onDefect(cause)
-        case None =>
-          update(a).foldCauseM(c => onDefect(c).ensuring(promise.halt(c)), {
-            case (b, a) => ref.set(a) <* promise.succeed(b)
-          })
-      }
-  }
-
-  /**
-   * Creates a new `RefM` with the specified value.
-   *
-   * @param a value of the new `RefM`
-   * @param n requested capacity for the bounded queue
-   * @param onDefect function that handles defects
-   * @tparam A type of the value
-   * @return `UIO[RefM[A]]`
-   */
-  def make[A](
-    a: A,
-    n: Int = 1000,
-    onDefect: Cause[Any] => UIO[Unit] = _ => IO.unit
-  ): UIO[RefM[A]] =
-    for {
-      ref   <- Ref.make(a)
-      queue <- Queue.bounded[Bundle[_, A, _]](n)
-      _     <- queue.take.flatMap(b => ref.get.flatMap(a => b.run(a, ref, onDefect))).forever.forkDaemon
-    } yield new RefM[A](ref, queue)
-
+  def signalRef[A](a: A): UIO[(RefM[A], Queue[A])] =
```

Comment:
A few suggestions:
- The queue should be a `Dequeue[A]` - not sure if we added this here or in `feature/zconduit`, but it's ZQueue for which the input is Nothing.
- Perhaps we could call this `subscriptionRef`, and have it return something like:
```scala
trait SubscriptionRefM[A] extends RefM[A] { // or ZRef[... Any, A]
  def await: UIO[A] = queue.take
}

def subscriptionRef[A](a: A): UIO[SubscriptionRefM[A]]
```
- For the variant in `ZStream`, we could have another subtype that extends `SubscriptionRefM` with `def changes = ZStream.fromQueue(queue)`.

Reply:
@iravid I updated to return a `Dequeue` and to change the name. I actually had that type signature in an earlier version and thought the type signature made it a little intimidating, but `Dequeue` is a great addition!

I don't think it is as straightforward to extend `RefM` because it is a sealed trait and we rely on that for the exhaustiveness of pattern matching in the implementation. Also I would think if we exposed a `SubscriptionRef` trait we would want the different combinators to return a `SubscriptionRef` so would have to override them.

This goes a little bit to whether we think of `Ref` as a data type or an interface. Right now `Ref` and `RefM` are very much data types so it isn't really possible to extend them, but you could potentially see them more as interfaces, potentially with `RefM` and `TRef` extending `Ref`. That could allow you to substitute different `Ref` implementations or extend them. Not sure if that makes sense.

Reply:
Right, forgot about that. What do you think about this for the streams module:
```
case class SubscriptionRef(ref: RefM[A], changes: ZStream[A])
```
?

Reply:
That can also have FunctorFilter methods later on.

Reply:
I think that is super straightforward. Wraps it up into a nice pair. Very easy to use. And the user can easily access the underlying `Ref` if they want to, for example to swap into a section of their code. Makes me wonder if we should do something similar for core.

Reply:
Updated to add `SubscriptionRef` and change name of combinator on `RefM` to `dequeueRef`.